// angular-instantsearch/esm2015/configure/configure.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Inject, forwardRef } from "@angular/core";
import { connectConfigure } from "instantsearch.js/es/connectors";
import { BaseWidget } from "../base-widget";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";
import { noop } from "../utils";
export class NgAisConfigure extends BaseWidget {
    /**
     * @param {?} instantSearchParent
     */
    constructor(instantSearchParent) {
        super("Configure");
        this.instantSearchParent = instantSearchParent;
        this.searchParameters = {};
        this.state = {
            refine: noop
        };
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.createWidget(connectConfigure, {
            searchParameters: this.searchParameters
        });
        super.ngOnInit();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (!changes["searchParameters"].isFirstChange) {
            this.state.refine(changes["searchParameters"].currentValue);
        }
    }
}
NgAisConfigure.decorators = [
    { type: Component, args: [{
                selector: "ais-configure",
                template: ""
            },] },
];
/** @nocollapse */
NgAisConfigure.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [forwardRef(() => NgAisInstantSearch),] }] }
];
NgAisConfigure.propDecorators = {
    searchParameters: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgAisConfigure.prototype.searchParameters;
    /** @type {?} */
    NgAisConfigure.prototype.state;
    /** @type {?} */
    NgAisConfigure.prototype.instantSearchParent;
}
