// angular-instantsearch/esm5/instantsearch/instantsearch.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { HttpClientModule } from "@angular/common/http";
import { NgAisInstantSearch } from "./instantsearch";
var NgAisInstantSearchModule = /** @class */ (function () {
    function NgAisInstantSearchModule() {
    }
    /**
     * @return {?}
     */
    NgAisInstantSearchModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgAisInstantSearchModule,
            providers: []
        };
    };
    NgAisInstantSearchModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [NgAisInstantSearch],
                    entryComponents: [NgAisInstantSearch],
                    exports: [NgAisInstantSearch],
                    imports: [CommonModule, HttpClientModule]
                },] },
    ];
    return NgAisInstantSearchModule;
}());
export { NgAisInstantSearchModule };
