// angular-instantsearch/esm5/configure/configure.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, Inject, forwardRef } from "@angular/core";
import { connectConfigure } from "instantsearch.js/es/connectors";
import { BaseWidget } from "../base-widget";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";
import { noop } from "../utils";
var NgAisConfigure = /** @class */ (function (_super) {
    tslib_1.__extends(NgAisConfigure, _super);
    function NgAisConfigure(instantSearchParent) {
        var _this = _super.call(this, "Configure") || this;
        _this.instantSearchParent = instantSearchParent;
        _this.searchParameters = {};
        _this.state = {
            refine: noop
        };
        return _this;
    }
    /**
     * @return {?}
     */
    NgAisConfigure.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.createWidget(connectConfigure, {
            searchParameters: this.searchParameters
        });
        _super.prototype.ngOnInit.call(this);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    NgAisConfigure.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (!changes["searchParameters"].isFirstChange) {
            this.state.refine(changes["searchParameters"].currentValue);
        }
    };
    NgAisConfigure.decorators = [
        { type: Component, args: [{
                    selector: "ais-configure",
                    template: ""
                },] },
    ];
    /** @nocollapse */
    NgAisConfigure.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [forwardRef(function () { return NgAisInstantSearch; }),] }] }
    ]; };
    NgAisConfigure.propDecorators = {
        searchParameters: [{ type: Input }]
    };
    return NgAisConfigure;
}(BaseWidget));
export { NgAisConfigure };
if (false) {
    /** @type {?} */
    NgAisConfigure.prototype.searchParameters;
    /** @type {?} */
    NgAisConfigure.prototype.state;
    /** @type {?} */
    NgAisConfigure.prototype.instantSearchParent;
}
