// angular-libphonenumber/pakmanaged.js
var global = Function("return this;")();
/*!
  * Ender: open module JavaScript framework (client-lib)
  * copyright Dustin Diaz & Jacob Thornton 2011 (@ded @fat)
  * http://ender.no.de
  * License MIT
  */
!function (context) {

  // a global object for node.js module compatiblity
  // ============================================

  context['global'] = context

  // Implements simple module system
  // losely based on CommonJS Modules spec v1.1.1
  // ============================================

  var modules = {}
    , old = context.$

  function require (identifier) {
    // modules can be required from ender's build system, or found on the window
    var module = modules[identifier] || window[identifier]
    if (!module) throw new Error("Requested module '" + identifier + "' has not been defined.")
    return module
  }

  function provide (name, what) {
    return (modules[name] = what)
  }

  context['provide'] = provide
  context['require'] = require

  function aug(o, o2) {
    for (var k in o2) k != 'noConflict' && k != '_VERSION' && (o[k] = o2[k])
    return o
  }

  function boosh(s, r, els) {
    // string || node || nodelist || window
    if (typeof s == 'string' || s.nodeName || (s.length && 'item' in s) || s == window) {
      els = ender._select(s, r)
      els.selector = s
    } else els = isFinite(s.length) ? s : [s]
    return aug(els, boosh)
  }

  function ender(s, r) {
    return boosh(s, r)
  }

  aug(ender, {
      _VERSION: '0.3.6'
    , fn: boosh // for easy compat to jQuery plugins
    , ender: function (o, chain) {
        aug(chain ? boosh : ender, o)
      }
    , _select: function (s, r) {
        return (r || document).querySelectorAll(s)
      }
  })

  aug(boosh, {
    forEach: function (fn, scope, i) {
      // opt out of native forEach so we can intentionally call our own scope
      // defaulting to the current item and be able to return self
      for (i = 0, l = this.length; i < l; ++i) i in this && fn.call(scope || this[i], this[i], i, this)
      // return self for chaining
      return this
    },
    $: ender // handy reference to self
  })

  ender.noConflict = function () {
    context.$ = old
    return this
  }

  if (typeof module !== 'undefined' && module.exports) module.exports = ender
  // use subscript notation as extern for Closure compilation
  context['ender'] = context['$'] = context['ender'] || ender

}(this);
// pakmanager:angular-libphonenumber
(function (context) {
  
  var module = { exports: {} }, exports = module.exports
    , $ = require("ender")
    ;
  
  /**
     * angular-libphonenumber
     * Nathan Hammond's libphonenumber ported to an angular filter
     * @version v1.0.0
     * @link https://github.com/cwill747/angular-libphonenumber
     * @license Apache-2.0
     */
    !function(e){e.module("cwill747.phonenumber",[]).directive("phoneNumber",["$log",function(e){function t(e){return e?e.replace(/([^0-9|+])/g,""):e}function n(t,n){var r=t;try{r=phoneUtils.formatAsTyped(t,n)}catch(o){e.debug(o)}return r}return{restrict:"A",require:"?ngModel",scope:{countryCode:"="},controllerAs:"",controller:function(){this.countryCode=this.countryCode||"us"},link:function(r,o,u,i){function l(e){return i.$isEmpty(e)?e:n(t(e),r.countryCode)}function c(e){if(i.$isEmpty(e))return e;var n=s(e);if(n===e)return e;var r=h.selectionStart,o=h.selectionEnd+n.length-e.length;return e.length<n.length&&(r+=n.length-e.length),e.length>n.length+1&&(r-=n.length-e.length),i.$setViewValue(n),i.$render(),h.setSelectionRange(r,o),t(n)}function s(e){var o=t(e),u="";return u=o&&o.length>1?n(o,r.countryCode):o,u.trim()}function a(t){var n=!1;try{n=phoneUtils.isValidNumberForRegion(t,r.countryCode)}catch(o){e.debug(o)}var u=i.$isEmpty(t)||n;return i.$setValidity("phoneNumber",u),t}var h=o[0];r.$watch("countryCode",function(){i.$modelValue=i.$viewValue+" "}),i.$formatters.push(l),i.$formatters.push(a),i.$parsers.push(c),i.$parsers.push(a)}}}])}(angular);
  provide("angular-libphonenumber", module.exports);
}(global));
