// angular-instantsearch/esm2015/angular-instantsearch.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { createSSRAlgoliaClient, createSSRSearchClient, parseServerRequest, BaseWidget, NgAisInstantSearch, NgAisRootModule, NgAisModule } from './index';
export { NgAisBreadcrumb as ɵbo } from './breadcrumb/breadcrumb';
export { NgAisBreadcrumbModule as ɵbn } from './breadcrumb/breadcrumb.module';
export { NgAisClearRefinements as ɵi } from './clear-refinements/clear-refinements';
export { NgAisClearRefinementsModule as ɵh } from './clear-refinements/clear-refinements.module';
export { NgAisConfigure as ɵbu } from './configure/configure';
export { NgAisConfigureModule as ɵbt } from './configure/configure.module';
export { NgAisCurrentRefinements as ɵbf } from './current-refinements/current-refinements';
export { NgAisCurrentRefinementsModule as ɵbe } from './current-refinements/current-refinements.module';
export { NgAisHierarchicalMenu as ɵbh } from './hierarchical-menu/hierarchical-menu';
export { NgAisHierarchicalMenuItem as ɵbi } from './hierarchical-menu/hierarchical-menu-item';
export { NgAisHierarchicalMenuModule as ɵbg } from './hierarchical-menu/hierarchical-menu.module';
export { NgAisHighlight as ɵe } from './highlight/highlight';
export { NgAisHighlightModule as ɵd } from './highlight/highlight.module';
export { NgAisHitsPerPage as ɵr } from './hits-per-page/hits-per-page';
export { NgAisHitsPerPageModule as ɵq } from './hits-per-page/hits-per-page.module';
export { NgAisHits as ɵc } from './hits/hits';
export { NgAisHitsModule as ɵb } from './hits/hits.module';
export { NgAisInfiniteHits as ɵbd } from './infinite-hits/infinite-hits';
export { NgAisInfiniteHitsModule as ɵbc } from './infinite-hits/infinite-hits.module';
export { NgAisInstantSearchModule as ɵa } from './instantsearch/instantsearch.module';
export { NgAisMenu as ɵk } from './menu/menu';
export { NgAisMenuModule as ɵj } from './menu/menu.module';
export { NgAisNumericMenu as ɵx } from './numeric-menu/numeric-menu';
export { NgAisNumericMenuModule as ɵw } from './numeric-menu/numeric-menu.module';
export { NgAisNumericSelector as ɵv } from './numeric-selector/numeric-selector';
export { NgAisNumericSelectorModule as ɵu } from './numeric-selector/numeric-selector.module';
export { NgAisPagination as ɵm } from './pagination/pagination';
export { NgAisPaginationModule as ɵl } from './pagination/pagination.module';
export { NgAisPanel as ɵbs } from './panel/panel';
export { NgAisPanelModule as ɵbr } from './panel/panel.module';
export { NgAisRangeInput as ɵbq } from './range-input/range-input';
export { NgAisRangeInputModule as ɵbp } from './range-input/range-input.module';
export { NgAisRangeSlider as ɵbm } from './range-slider/range-slider';
export { NgAisRangeSliderModule as ɵbl } from './range-slider/range-slider.module';
export { NgAisRatingMenu as ɵbk } from './rating-menu/rating-menu';
export { NgAisRatingMenuModule as ɵbj } from './rating-menu/rating-menu.module';
export { NgAisFacetsSearch as ɵp } from './refinement-list/facets-search';
export { NgAisRefinementList as ɵo } from './refinement-list/refinement-list';
export { NgAisRefinementListModule as ɵn } from './refinement-list/refinement-list.module';
export { NgAisSearchBox as ɵg } from './search-box/search-box';
export { NgAisSearchBoxModule as ɵf } from './search-box/search-box.module';
export { NgAisSortBy as ɵt } from './sort-by/sort-by';
export { NgAisSortByModule as ɵs } from './sort-by/sort-by.module';
export { NgAisStats as ɵz } from './stats/stats';
export { NgAisStatsModule as ɵy } from './stats/stats.module';
export { NgAisToggle as ɵbb } from './toggle/toggle';
export { NgAisToggleModule as ɵba } from './toggle/toggle.module';
