// angular-instantsearch/esm2015/numeric-selector/numeric-selector.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, Inject, forwardRef } from "@angular/core";
import { connectNumericSelector } from "instantsearch.js/es/connectors";
import { BaseWidget } from "../base-widget";
import { NgAisInstantSearch } from "../instantsearch/instantsearch";
import { noop } from "../utils";
/** @typedef {?} */
var NumericSelectorState;
export { NumericSelectorState };
export class NgAisNumericSelector extends BaseWidget {
    /**
     * @param {?} instantSearchParent
     */
    constructor(instantSearchParent) {
        super("NumericSelector");
        this.instantSearchParent = instantSearchParent;
        this.operator = "=";
        this.state = {
            currentRefinement: null,
            options: [],
            refine: noop
        };
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.createWidget(connectNumericSelector, {
            attributeName: this.attribute,
            operator: this.operator,
            options: this.items
        });
        super.ngOnInit();
    }
}
NgAisNumericSelector.decorators = [
    { type: Component, args: [{
                selector: "ais-numeric-selector",
                template: `
    <div [class]="cx('')">
      <select
        [class]="cx('select')"
        (change)="state.refine($event.target.value)"
      >
        <option
          [class]="cx('option')"
          *ngFor="let item of state.options"
          [value]="item.value"
          [selected]="item.value === state.currentRefinement"
        >
          {{item.label}}
        </option>
      </select>
    </div>
  `
            },] },
];
/** @nocollapse */
NgAisNumericSelector.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [forwardRef(() => NgAisInstantSearch),] }] }
];
NgAisNumericSelector.propDecorators = {
    attribute: [{ type: Input }],
    operator: [{ type: Input }],
    items: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    NgAisNumericSelector.prototype.attribute;
    /** @type {?} */
    NgAisNumericSelector.prototype.operator;
    /** @type {?} */
    NgAisNumericSelector.prototype.items;
    /** @type {?} */
    NgAisNumericSelector.prototype.state;
    /** @type {?} */
    NgAisNumericSelector.prototype.instantSearchParent;
}
